slycot/math.py

```diff
@@ -70,6 +70,361 @@ def mc01td(dico,dp,p):
     return out[:-2]
 
 
+def mb03vd(n, ilo, ihi, A):
+    """ HQ, Tau = mb03vd(n, ilo, ihi, A)
+
+    To reduce a product of p real general matrices A = A_1*A_2*...*A_p
+    to upper Hessenberg form, H = H_1*H_2*...*H_p, where H_1 is
+    upper Hessenberg, and H_2, ..., H_p are upper triangular, by using
+    orthogonal similarity transformations on A,
+
+            Q_1' * A_1 * Q_2 = H_1,
+            Q_2' * A_2 * Q_3 = H_2,
+                   ...
+            Q_p' * A_p * Q_1 = H_p.
+
+    Parameters
+    ----------
+
+    n : int
+            The order of the square matrices A_1, A_2, ..., A_p.
+            n >= 0.
+
+    ilo, ihi : int
+            It is assumed that all matrices A_j, j = 2, ..., p, are
+            already upper triangular in rows and columns [:ilo] and
+            [ihi:n], and A_1 is upper Hessenberg in rows and columns
+            [:ilo] and [ihi:n], with A_1[ilo-1,ilo] = 0 (unless
+            ilo = 1), and A_1[ihi,ihi-1] = 0 (unless ihi = n).
+            If this is not the case, ilo and ihi should be set to 1
+            and n, respectively.
+            1 <= ilo <= max(1,n); min(ilo,n) <= ihi <= n.
+
+    A : ndarray
+            A[:n,:n,:p] must contain the matrices of factors to be reduced;
+            specifically, A[:,:,j-1] must contain A_j, j = 1, ..., p.
+
+
+    Returns
+    -------
+
+    HQ : ndarray
```

Comment:
I think specifying the exact size of `HQ` would be good (I gather `HQ.shape` will be `(n, n, p)` ?

Similarly, could you say what the size of `Tau` will be?  (this one I couldn't guess)

Reply:
`HQ.shape` is the same shape as input A. Minimum is `(max(n, 1), max(n, 1), p)` but could be larger.
`Tau.shape` is `(max(1, n-1), p)`

Reply:
Isn't it worth adding this to the docstring?

If `n < min(A.shape[:2])`, `HQ` would still have the same size as A?  Would the `n:` rows and columns of HQ then be uninitialized?

Reply:
Yes that's the way many of the SLICOT routines do it. Adding this to the docstring and rebasing once more.

We have a bunch of different docstring styles for the various functions. Opening another issue for that.

Reply:
Thank you.  Do you agree with this, from https://github.com/python-control/Slycot/pull/88#discussion_r407025183:

     So should that be "upper diagonal (or Hessenberg) in [:ilo-1]" ?

Reply:
Agree. General rule is:
- `Python-index = Fortran-index - 1`
- `Python-slice-start = Fortran-slice-start-1 
- `Python-slice-end:Fortran-slice-end`

Or put differently:
` A(x,y) = A[x-1,y-1]`
` A(a:b,c:d) = A[a-1:b,c-1:d]`

Reply:
Fix committed